### PR TITLE
Moved super.dispose down

### LINKED ISSUE #802: Dispose order is wrong
**Describe the bug**
The order of the dispose method is wrong. 
Now, it first disposes the object and afterwards tries to remove the listeners, which doesn't make sense since it is already disposed. A package I use also throws an error if you execute code after disposing, which does make sense. 

[Flutter specification on state dispose](https://api.flutter.dev/flutter/widgets/State/dispose.html) is relevant here: Implementations of this method should **end** with a call to the inherited method, as in super.dispose().
I would strongly recommend adding this to the documentation and handling it the same. 

In short this:
```dart
  @override
  void dispose() {
    super.dispose();
    _removeListener?.call();
    if (_didInitValue) {
      delegate.dispose?.call(element!, _value as T);
    }
  }
```
should become this:
```dart
  @override
  void dispose() {
    _removeListener?.call();
    if (_didInitValue) {
      delegate.dispose?.call(element!, _value as T);
    }
    super.dispose();
  }
```

### PR DESCRIPTION
**Pull request by ikbendewilliam**
Fixes #802

### CODE REVIEW

**Comment by rrousselGit** [comment]
Hello!

Could you write tests for these changes?

**Comment by ikbendewilliam** [comment]
I'm having trouble writing a good test, I can't seem to recreate the same issue as we have in our plugin. I did add a test that validates the delegate isn't disposed before the listeners have been removed though 🤔

File: test/null_safe/change_notifier_provider_test.dart [hunk 1]
```diff
@@ -356,4 +356,61 @@ void main() {
...
+      builder: (context, __) => Text(
+        Provider.of<_ChangeNotifierDisposeTest>(context).text,
+        textDirection: TextDirection.ltr,
+      ),
+    );
+
+    await tester.pumpWidget(widget);
+    await tester.pumpAndSettle();
+
+    // If the listener is removed after the super.dispose, this will throw
```
Comment (rrousselGit) [inline test/null_safe/change_notifier_provider_test.dart:373]: That's not the case. ChangeNotfier no-longer throws if removeListener is invoked post dispose.

The assert has been removed.

Comment (rrousselGit) [reply test/null_safe/change_notifier_provider_test.dart:373]: Sorry by bad, I had yet to read the custom exception bit

**Comment by rrousselGit** [comment]
@ikbendewilliam Did these changes fix the issue with your plugin?

**Review by rrousselGit — APPROVED** [review]
LGTM, but make sure it works on your app first. It'd be a shame if it didn't

**Comment by ikbendewilliam** [comment]
@rrousselGit Turns out I'm sick, so when I'm feeling better I'll have another go at writing better tests and I'll confirm that everything is fixed for our plugin. Thank you for your patience

**Comment by rrousselGit** [comment]
No problem! Thanks for your work :)

**Comment by ikbendewilliam** [comment]
I have good news and bad news 😅 
The good news is that I'm able to reproduce it. The bad news is that I used two providers and the error I was receiving was not because the order of the dispose is wrong, but because `dispose` got called a second time. 
This branch obviously doesn't fix that since the issue is not in the package. 

My opinion, however, is still that the order should be first remove listeners and then super.dispose. So for me this can be merged, WDYT?

**Comment by rrousselGit** [comment]
Honestly I'm a bit skeptical.
This could be breaking some folks, since that behavior has been around for ages now.

And the current behavior isn't problematic. You're the first to point it out, I never saw/heard of the issue before.
Removing the listeners _after_ dispose is generally harmless for most provided objects.

**Comment by rrousselGit** [comment]
Hello again!

I don't think I will go through with these changes after-all. In the end:
- provider is in Long Term Maintenance mode. I avoid to avoid breaking changes
- this could be silently breaking
- the value is not that large
- people haven't requested for this before.

So I'll keep things as they are for now.
But this could be revisited if a large number of people ask for this


I appreciate your effort though!

**Comment by snaeji** [comment]
We are in the same situation. It would be very beneficial for us if this would be merged so we don't have to fork 👍

**Pull request closed without merging**